pythainlp/translate/core.py

```diff
@@ -16,121 +16,132 @@
 
 
-download_model()
+model = None
+model_name = None
 
 
-def get_path(model, path1, path2, file= None) -> str:
+def get_path(model, path1, path2, file=None) -> str:
     path = os.path.join(os.path.join(get_full_data_path(model), path1), path2)
     if file is not None:
         return os.path.join(path, file)
     return os.path.join(path, "")
 
 
-en2th_word2bpe_model = TransformerModel.from_pretrained(
-    model_name_or_path=get_path(
-        "scb_1m_en-th_moses",
-        "SCB_1M+TBASE_en-th_moses-spm_130000-16000_v1.0",
-        "models",
-    ),
-    checkpoint_file="checkpoint.pt",
-    data_name_or_path=get_path(
-        "scb_1m_en-th_moses",
-        "SCB_1M+TBASE_en-th_moses-spm_130000-16000_v1.0",
-        "vocab",
-        ),
-)
+def en2th_word2bpe_model():
+    global model, model_name
+    download_install("scb_1m_en-th_moses")
+    if model_name != "en2th_word2bpe":
+        del model
+        model = TransformerModel.from_pretrained(
+            model_name_or_path=get_path(
+                "scb_1m_en-th_moses",
+                "SCB_1M+TBASE_en-th_moses-spm_130000-16000_v1.0",
+                "models",
+            ),
+            checkpoint_file="checkpoint.pt",
+            data_name_or_path=get_path(
+                "scb_1m_en-th_moses",
+                "SCB_1M+TBASE_en-th_moses-spm_130000-16000_v1.0",
+                "vocab",
+                ),
+        )
+        model_name = "en2th_word2bpe"
 
 
 def _en2th_word2bpe_translate(text: str) -> str:
+    global model, model_name
+    en2th_word2bpe_model()
     tokenized_sentence = " ".join(
         en_word_tokenize.tokenize(text)
     )
-    hypothesis = en2th_word2bpe_model .translate(tokenized_sentence)
+    hypothesis = model.translate(tokenized_sentence)
     hypothesis = hypothesis.replace(" ", "").replace("▁", " ")
     return hypothesis
 
 
-th2en_word2word_model = TransformerModel.from_pretrained(
-    model_name_or_path=get_path(
-        "scb_1m_th-en_newmm",
-        "SCB_1M+TBASE_th-en_newmm-moses_130000-130000_v1.0",
-        "models",
-    ),
-    checkpoint_file="checkpoint.pt",
-    data_name_or_path=get_path(
-        "scb_1m_th-en_newmm",
-        "SCB_1M+TBASE_th-en_newmm-moses_130000-130000_v1.0",
-        "vocab",
-    ),
-)
+def th2en_word2word_model():
+    global model, model_name
+    download_install("scb_1m_th-en_newmm")
+    if model_name != "th2en_word2word":
+        del model
+        model = TransformerModel.from_pretrained(
+            model_name_or_path=get_path(
+                "scb_1m_th-en_newmm",
+                "SCB_1M+TBASE_th-en_newmm-moses_130000-130000_v1.0",
+                "models",
+            ),
+            checkpoint_file="checkpoint.pt",
+            data_name_or_path=get_path(
+                "scb_1m_th-en_newmm",
+                "SCB_1M+TBASE_th-en_newmm-moses_130000-130000_v1.0",
+                "vocab",
+            ),
+        )
+        model_name = "th2en_word2word"
```

Comment:
It looks like we're implicitly using the Singleton pattern here.  I think it's might be better to do it in a bit more principled way.

For example, this is what I use in Attacut: https://github.com/PyThaiNLP/attacut/blob/master/attacut/tokenizer.py#L65